garden_ai/utils/misc.py

```diff
@@ -21,80 +16,6 @@
 issubtype = beartype.door.is_subhint
 
 
-def get_cache_tag(
-    pip_reqs: List[str], conda_deps: List[str], py_version: Optional[str]
-) -> str:
-    """Used for container uuid caching based on pip/conda dependencies"""
-    full_spec = (
-        pip_reqs + conda_deps + [py_version or ".".join(map(str, sys.version_info[:3]))]
-    )
-    return str(crc32(str(sorted(set(full_spec))).encode()))
-
-
-def safe_compose(f: Callable, g: Callable):
-    """Helper: compose function `f` with function `g`, provided their annotations indicate compatibility.
-    Arguments with defaults are ignored.
-
-    Parameters
-    ----------
-    f : Callable
-    g : Callable
-
-    Raises
-    ------
-    TypeError
-        If the annotations for `f`'s argument types and `g`'s return
-        type are not equivalent.
-    """
-
-    f_sig: Signature = signature(f)
-    g_sig: Signature = signature(g)
-
-    f_in = tuple(
-        p.annotation for p in f_sig.parameters.values() if p.default is Parameter.empty
-    )
-    g_out = g_sig.return_annotation
-
-    if len(f_in) == 1:
-        if issubtype(g_out, f_in[0]):
-            # note that we do NOT unpack g's output
-            def f_of_g(*args, **kwargs):
-                return f(g(*args, **kwargs))
-
-        else:
-            raise TypeError(
-                (
-                    f"Could not compose step {f.__name__} with step {g.__name__} "
-                    "due to return type signature mismatch: expected "
-                    f"{f_in[0]}, got {g_out}."
-                )
-            )
-    elif len(f_in) > 1:
-        raise TypeError(
-            (
-                f"Could not compose step {f.__name__} with step {g.__name__} "
-                f"{f.__name__} has more than 1 positional (required) argument, "
-                "but only 1 would be given."
-            )
-        )
-    else:
-        raise TypeError(
-            (
-                f"Could not compose {f.__name__} with step {g.__name__}. "
-                f"{f.__name__} takes 0 positional arguments, but would be called "
-                f"on {g.__name__}'s return value."
-            )
-        )
-    # give the returned function a new signature, corresponding
-    # to g's input types and f's return type
-    f_of_g.__signature__ = Signature(  # type: ignore
-        parameters=list(g_sig.parameters.values()),
-        return_annotation=f_sig.return_annotation,
-    )
-    f_of_g.__name__ = f.__name__ + "_COMPOSED_WITH_" + g.__name__
-    return f_of_g
-
-
 def garden_json_encoder(obj):
```

Comment:
I think this one can go too! Since we're not trying to serialize function types to json anymore
```suggestion

```

Reply:
It looks like we hit

```
E       TypeError: Object of type UUID is not JSON serializable
```

in the unit tests if I try to take this out of the local DB operations. So this accidentally got UUID serialization working. Not gonna tackle it here but let's try and see if we can mop it up later.